File: client/src/main/java/com/fiwio/iot/demeter/app/demeter/splash/SplashActivity.java
package com.fiwio.iot.demeter.app.demeter.splash;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.RequiresApi;
import android.support.v7.app.AppCompatActivity;

import com.fiwio.iot.demeter.app.demeter.app.DemeterApplication;
import com.fiwio.iot.demeter.app.demeter.discovery.DemerServiceFound;
import com.fiwio.iot.demeter.app.demeter.discovery.MulticastDns;
import com.fiwio.iot.demeter.app.demeter.main.MainActivity;
import com.fiwio.iot.demeter.app.demeter.splash.di.SplashModule;
import com.fiwo.iot.demeter.app.R;

import javax.inject.Inject;

public class SplashActivity extends AppCompatActivity implements DemerServiceFound {

    public static final int TWENTY_SECONDS = 5000;
    @Inject
    MulticastDns multicastDns;

    private Handler mHandler = new Handler();
    private int repeatCount = 0;
    private boolean started = false;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_splash);
        inject();
    }

    private void inject() {
        DemeterApplication.get(this).getAppComponent().plus(new SplashModule()).injects(this);
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        if (multicastDns != null) {
            multicastDns.discoverServices(this, mHandler);
            startWatchDog();
        } else {
            onServiceSearchFailed();
        }
    }

    private void startWatchDog() {
        new Thread(new Runnable() {


            @Override
            public void run() {
                try {
                    Thread.sleep(TWENTY_SECONDS);
                    mHandler.post(new Runnable() {
                        @Override
                        public void run() {
                            onServiceSearchFailed();
                        }
                    });
                } catch (Exception e) {
                    ;
                }
            }
        }).start();
    }


    @Override
    public void onServiceFound(String ipAddress) {
        multicastDns.stopDiscovery();
        multicastDns = null;
        Intent mainIntent = new Intent(this, MainActivity.class);
        Bundle bundle = new Bundle();
        bundle.putString("host", ipAddress);
        mainIntent.putExtras(bundle);
        this.startActivity(mainIntent);
        this.finish();
    }

    @Override
    public void onServiceSearchFailed() {
        if (repeatCount >= 3) {
            multicastDns.stopDiscovery();
            if (!started) {
                started = true;
                Intent mainIntent = new Intent(this, MainActivity.class);
                this.startActivity(mainIntent);
                this.finish();
            }
        } else {
            repeatCount++;
            if (multicastDns != null) {
                multicastDns.stopDiscovery();
                multicastDns.discoverServices(this, mHandler);
                startWatchDog();
            }

        }
    }
}

File: mobile-ui/src/main/java/com/fiwio/iot/demeter/android/ui/feature/splash/SplashActivity.kt
package com.fiwio.iot.demeter.android.ui.feature.splash

import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.support.v7.app.AppCompatActivity
import android.view.View
import com.fiwio.iot.demeter.android.ui.R
import com.fiwio.iot.demeter.android.ui.ext.getAppComponent
import com.fiwio.iot.demeter.android.ui.feature.main.MainActivity
import com.fiwio.iot.demeter.android.ui.feature.main.MainNavigator
import com.fiwio.iot.demeter.android.ui.feature.splash.di.SplashComponent
import com.fiwio.iot.demeter.android.ui.feature.splash.di.SplashModule
import com.fiwio.iot.demeter.presentation.feature.splash.SplashContract
import com.fiwio.iot.demeter.presentation.model.DnsLookupState
import kotlinx.android.synthetic.main.activity_splash.*
import javax.inject.Inject


class SplashActivity : AppCompatActivity(), SplashContract.View, MainNavigator {

    override fun showContent() {
        loading_container.visibility = View.VISIBLE
    }

    override fun setData(data: DnsLookupState) {
        if (data.found) {
            navigateToMainWithUrl(data.ipAdress)
        }
    }


    @Inject
    internal lateinit var presenter: SplashContract.Presenter

    private lateinit var component: SplashComponent

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash)

        component = getAppComponent().plus(SplashModule())
        component.injects(this)
    }


    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        presenter.attachView(this)

        manual_ip_button.setOnClickListener {
            val text = manual_ip_edit.editableText.toString()
            navigateToMainWithUrl(text)
        }

        start.setOnClickListener {
            val text = manual_ip_edit.editableText.toString()
            navigateToMainWithUrl("http://192.168.3.54:8080")
        }
    }

    override fun enterUrlByHand() {
        loading_container.visibility = View.GONE
        manual_ip_container.visibility = View.VISIBLE
    }

    override fun showLoading(pullToRefresh: Boolean) {
        loading_container.visibility = View.VISIBLE
        manual_ip_container.visibility = View.GONE
    }

    override fun showError(e: Throwable, pullToRefresh: Boolean) {
    }

    override fun loadData(pullToRefresh: Boolean) {
    }

    override fun navigateToMainWithUrl(url: String) {
        MainActivity.navigate(this, url)
    }

    override fun getSystemService(name: String?): Any {
        return when (name) {
            "component" -> component
            else -> super.getSystemService(name)
        }
    }
}
